---
title: "R markdown, summary statistics and reproducible research (STAT5003)"
author: "Pengyi Yang"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  code_folding: show
  fig_height: 10
  fig_width: 10
  number_sections: yes
  theme: united
  toc: TRUE
---

> ## Outlines {.objectives}
> * R data types.
> * R markdown.
> * Review basic statistical concept in data analysis.
> * Reproducible research - using R markdown to generate report.

## Introduction

### Learning objective
The aim here is to illustrate how to use R markdown to create data analysis report in a reproduciable manner. We will also review several summary statistics for exploratory data analysis and visualization.


## R data types

#### Atomic data types and operations
**Character**
```{r}
species <- "mouse"
```
**Integer**
```{r}
sampleSize <- 6
```
**Numeric**
```{r}
meanLifeSpan <- 2.05
```
**Factor**
```{r}
sex <- factor("male")
```
**Boolean**
```{r}
mammal <- TRUE
```
**Combine characters**
```{r}
strain <- "C57BL/6"
sampleDetail <- paste(species, strain)
sampleDetail
```


#### Primitive data types and operations
**Vectors**
```{r}
Groups <- c("treated", "control", "control", "control", "treated", "treated")
Groups
lifeSpans <- c(2.1, 1.9, 1.7, 1.8, 2.3, 2.5)
lifeSpans
# Calculate mean life span of sample population
mean(lifeSpans)
```
**List**
```{r}
# Split samples based on their groups
sampleByGroup <- split(lifeSpans, Groups)
sampleByGroup # sampleByGroup is a list
# access a single element in a list (get the mean of the second group)
sampleByGroup$treated # access by name using '$'
sampleByGroup[[2]] # access by index using '[[]]'
# calculate mean of samples in each group
means <- lapply(sampleByGroup, mean)
means
# calculate standard deviation of samples in each group
sds <- lapply(sampleByGroup, sd)
sds
sds[[2]] # access list by index
```
**Matrix**
```{r}
# using cbind (column bind) to combine two vectors into a matrix
mat <- cbind(sampleByGroup$treated, sampleByGroup$control)
mat
colnames(mat) <- c("treated", "control") # assign column names
mat
rownames(mat) <- paste("sample", 1:nrow(mat)) # assgin row names
mat
```
**Data frame**
```{r}
# create a data frame where life spans are stored in the first column and grouping information are stored in the second column
df <- data.frame(lifeSpans, Groups)
df
# convert data frame into a list using function 'split'
sampleByGroup <- split(df[,1], df[,2])
sampleByGroup
# visualise the data using boxplot
boxplot(sampleByGroup$control, sampleByGroup$treated, ylab="Life span (year)")
# perform a t-test with respect to life span between the two groups of samples
g1 <- sampleByGroup$treated
g2 <- sampleByGroup$control
t.test(g1, g2, alternative="greater")
```


## Libraries to load
```{r loading, warning=FALSE, message=FALSE}
library(reshape2)
library(ggplot2)
library(devtools)
#install_github("vqv/ggbiplot") # run this line if you haven't installed ggbiplot
library(ggbiplot)
library(gplots)
library(gridExtra)
library(scales)
library(limma)
```



### Iris dataset description
This famous (Fisher's or Anderson's) iris data set gives the measurements in centimeters of the variables sepal length and width and petal length and width, respectively, for 50 flowers from each of 3 species of iris. The species are Iris setosa, versicolor, and virginica.

References
Becker, R. A., Chambers, J. M. and Wilks, A. R. (1988) The New S Language. Wadsworth & Brooks/Cole. (has iris3 as iris.)

## Data exploration on iris dataset
```{r, warning=FALSE, message=FALSE}
# obtain and summarize phenodata
dim(iris)
head(iris)
table(iris$Species)
class(iris)
```

Let us first extract data from a single variable "Sepal Length". We will use this to review basic R commands for numeric and graphical summaries.

### Numerical summaries for univariate data
```{r, warning=FALSE, message=FALSE}
sepalLength <- iris$Sepal.Length
mean(sepalLength) ## Calculating the mean 
median(sepalLength) ## Calculating the median
quantile(sepalLength) ## Calculating quantiles
IQR(sepalLength) ##  Calculating the Interquartile range
sd(sepalLength) ## Calculating standard deviation
var(sepalLength) ## Calculating variance
range(sepalLength) ## Calculating the range 
max(sepalLength) ## Calculating maximum value 
min(sepalLength) ## Calculating minimum value 
summary(sepalLength) ## Numerical summary of "sepalLength"
```

### Graphical summaries for univariate data
```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=8}
par(mfrow=c(2, 2))
boxplot(sepalLength)
hist(sepalLength, breaks=10) ## Shows the distribution of expression of different genes by frequency
hist(sepalLength, breaks=10, probability = TRUE) ## Shows the distribution of expression of different genes by probability (density)
lines(density(sepalLength, adjust=2), lty="dotted", col="red", lwd=2) ## Estimate the density of sepalLength
```

### Numeric and graphical summary for bivariate data
Scatterplots provide insight into relationships between variables, for example, sugars and calories or fat and calories, etc.  You can also calculate correlation between any pairs of variables (as shown below).
```{r}
petalLength <- iris$Petal.Length
cor(sepalLength, petalLength) # correlation between sepal length and Petal length
lm(sepalLength~petalLength) # fitting a linear model to the two variables
coef(lm(sepalLength ~ petalLength)) # extract coefficients 
```

```{r, warning=FALSE, message=FALSE}
plot1 <- qplot(petalLength, sepalLength) + geom_abline(intercept = 4.3, slope = 0.409, col="blue", linetype = 2) # plot one variable vs another; correlation plot; add a regression line using intercept and slope obtained from the coef() function
plot2 <- qplot(petalLength, sepalLength) + geom_smooth(method = "lm", se = FALSE, col="red") ## Add a regression line using geom_smooth function
grid.arrange(plot1, plot2, ncol=2) # arrange plots using grid
```

### Species based analysis (split data, analyse and compare three species)
In this section, we review how to parse data and visualise them based on factors (species) of interest

```{r}
boxplot(split(sepalLength, iris$Species), ylab="Sodium", xlab="Species") # using base plot system. We are splitting the "sepal length" by using "species" as a factor here.
ggplot(iris, aes(factor(Species), Sepal.Length)) + geom_boxplot() # using ggplot system. Function "aes" are used to split the data
ggplot(iris, aes(y=Sepal.Length, x=Petal.Length)) + geom_point() + facet_grid(. ~ Species) + geom_smooth(method = "lm", se = TRUE, col="red")
```

## Data analysis on Leukemia gene expression dataset
### Leukemia dataset description
ExpressionSet containing gene expresion data from 60 bone marrow samples of patients with one
of the four main types of leukemia (ALL, AML, CLL, CML) or no-leukemia controls.

Platform: Affymetrix Human Genome U133 Plus 2.0  

Annotation: genemapperhgu133plus2 (CDF from GATExplorer)

Mapping: Gene Ensembl ID (20172 features)

Tissue: Bone Marrow

Cell type: Mononuclear cells isolated by Ficoll density centrifugation

Disease type:  
1. Acute Lymphoblastic Leukemia (ALL). Subtype: c-ALL / pre-B-ALL without t(9;22)  
2. Acute Myeloid Leukemia (AML). Subtype: Normal karyotype  
3. Chronic Lymphocytic Leukemia (CLL)  
4. Chronic Myeloid Leukemia (CML)  
5. Non-leukemia and healthy bone marrow (NoL)  

All samples were obtained from untreated patients at the time of diagnosis.
Preprocessing: The microarrays were normalized with RMA using a redefined probe mapping from
Affymetrix probesets to Ensembl genes (Ensembl IDs ENSG). This alternative Chip Definition
File (CDF) with complete unambiguous mapping of microarray probes to genes (GeneMapper)
is available at GATExplorer (http://bioinfow.dep.usal.es/xgate/mapping/mapping.php) (Risueno et
al. 2010).


### Aims
identify differentially expressed (DE) genes in two separate microarray datasets from two different species with similar experimental design and compare the results. We will look into mouse muscle gene expression (GE) data post high fat feeding (HFF) and compare it to GE of mouse fed a normal chow diet. In parallel, we will practice using various analysis steps on human muscle gene expression data of insulin resistant and insulin sensitive individuals. We will try to find the overlap between genes altered in mice under HFF and dysregulated genes in insulin resistant humans. 

#### obtain and load data
```{r, warning=FALSE, message=FALSE, eval=FALSE}
# download data from bioconductor 
source("https://bioconductor.org/biocLite.R")
biocLite("leukemiasEset")
```

```{r, warning=FALSE, message=FALSE}
# load data
library(leukemiasEset)
data(leukemiasEset)

# extract expression matrix from the "ExpressionSet object"
leukemia.mat <- exprs(leukemiasEset)
class(leukemia.mat)

# obtain and summarize phenodata
phenodata <- pData(leukemiasEset)
head(phenodata)
table(phenodata[,"LeukemiaType"])
```

#### Summarise data
```{r, warning=FALSE, message=FALSE}
dim(leukemia.mat)
head(leukemia.mat)
colnames(leukemia.mat)
```


## Hypothesis testing: 
### ANOVA for selecting differentially expressed genes
```{r, warning=FALSE, message=FALSE}
# extract phenotype data
f <- phenodata[colnames(leukemia.mat), 3]
# apply ANOVA to each row (aka each gene) of the leukemia matrix
leukemia.aov.p <- apply(leukemia.mat, 1, function(x){
	summary(aov(as.numeric(x) ~ f))[[1]][["Pr(>F)"]][1]
})
hist(leukemia.aov.p)
```

### Limma's moderated t-test and f-test
```{r, warning=FALSE, message=FALSE}
# create the design matrix
f = phenodata[colnames(leukemia.mat), 3]
design = model.matrix(~ f - 1)
colnames(design) <- c("ALL", "AML", "CLL", "CML", "NoL")
design
fit <- lmFit(leukemia.mat , design) # fitting data to a linear model
contrast.matrix <- makeContrasts(ALL-NoL, AML-NoL, CLL-NoL, CML-NoL, levels=design)  # defining group comparisons
fit2 <- contrasts.fit(fit, contrast.matrix) # convert the coefficients of the design matrix into these 4 contrasts which are to be tested equal to zero
fit2 <- eBayes(fit2) # With a series of related parameter estimates & standard errors, eBayes computes moderated t-statistics of differential expression by empirical Bayes shrinkage of the standard errors towards a common value.
topTable(fit2, coef=1, number=Inf)[1:10,] # Extract test results based on moderated t-test with respect to ALL
limma.f.p <- fit2$F.p.value # Extract the p-value from the moderated f-test model
names(limma.f.p) <- rownames(fit2)
```

### compare consensus of the two approaches
```{r, warning=FALSE, message=FALSE}
idx <- names(leukemia.aov.p)
plot(leukemia.aov.p[idx], limma.f.p[idx], col = alpha("black", 0.1))
```


### Note about this document
This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>. When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.  Basic session info can be generated using sessioninfo command.

# output session information
```{r, echo=FALSE}
sessionInfo()
```


